pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -166,46 +189,228 @@ func (pfactory *PredicateMetadataFactory) GetMetadata(pod *v1.Pod, nodeNameToInf
+		nodeTopologyMaps := newTopologyPairsMaps()
+		// nodeInfo.Pods() can be empty; or all pods don't fit
+		for _, existingPod := range nodeInfo.Pods() {
+			ok, err := podMatchesAllSpreadConstraints(existingPod, pod.Namespace, constraints)
+			if err != nil {
+				catchError(err)
+				cancel()
+				return
+			}
+			if ok {
+				for _, constraint := range constraints {
+					// constraint.TopologyKey is already guaranteed to be present
+					pair := topologyPair{key: constraint.TopologyKey, value: node.Labels[constraint.TopologyKey]}
+					nodeTopologyMaps.addTopologyPair(pair, existingPod)
+				}
+			}
+		}
+		// If needed, append "headless" topology pair
+		// For example, on node-x, there is no pod matching spread constraints
+		// but node-x should be also considered as a match (with match number 0)
+		// i.e. <node: node-x>: {}
+		for _, constraint := range constraints {
+			// constraint.TopologyKey is already guaranteed to be present
+			pair := topologyPair{
+				key:   constraint.TopologyKey,
+				value: node.Labels[constraint.TopologyKey],
+			}
+			// addTopologyPairWithoutPods is a non-op if other pods match this pair
+			nodeTopologyMaps.addTopologyPairWithoutPods(pair)
+		}
+
+		appendTopologyPairsMaps(nodeTopologyMaps)
+	}
+	workqueue.ParallelizeUntil(ctx, 16, len(allNodeNames), processNode)
+
+	if firstError != nil {
+		return nil, firstError
+	}
+
+	// calculate min match for each topology pair
+	topologyPairsPodSpreadMap.topologyKeyToMinPodsMap = make(map[string]int32, len(constraints))
+	for _, constraint := range constraints {
+		topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[constraint.TopologyKey] = math.MaxInt32
+	}
+	for pair, podSet := range topologyPairsPodSpreadMap.topologyPairToPods {
+		// TODO(Huang-Wei): short circuit all portions of <topologyKey: any value>
+		// if we see 0 as min match of the topologyKey
+		if l := int32(len(podSet)); l < topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key] {
+			topologyPairsPodSpreadMap.topologyKeyToMinPodsMap[pair.key] = l
+		}
+	}
+	return topologyPairsPodSpreadMap, nil
+}
+
+func getHardTopologySpreadConstraints(pod *v1.Pod) (constraints []v1.TopologySpreadConstraint) {
+	if pod != nil {
+		for _, constraint := range pod.Spec.TopologySpreadConstraints {
+			if constraint.WhenUnsatisfiable == v1.DoNotSchedule {
+				constraints = append(constraints, constraint)
+			}
+		}
+	}
+	return
+}
+
+func podMatchesAllSpreadConstraints(pod *v1.Pod, ns string, constraints []v1.TopologySpreadConstraint) (bool, error) {
+	if len(constraints) == 0 || pod.Namespace != ns {
+		return false, nil
+	}
+	return podLabelsMatchesSpreadConstraints(pod.Labels, constraints)
+}
+
+// some corner cases:
+// 1. podLabels = nil, constraint.LabelSelector = nil => returns false
+// 2. podLabels = nil => returns false
+// 3. constraint.LabelSelector = nil => returns false
+func podLabelsMatchesSpreadConstraints(podLabels map[string]string, constraints []v1.TopologySpreadConstraint) (bool, error) {
```

Comment:
It's to check if all constraints[*].labelSelector matches podLabels. Renaming the function to `spreadConstraintsMatchPodLabels` makes more sense?

Reply:
I'm not questioning what the function is doing, but rather why. In line 248, we are ignoring the pod if "all labels" don't match. I'm arguing that it should be "at least one label". AND vs OR.

Reply:
This is a good point Aldo. 

@Huang-Wei if two constraints have different label selectors, we should be computing the skew for each one separately, and so the min match may need to be computed per topology per "label selector", right?

Reply:
Oh, I understand your concerns now. This is a very good point and it's not the first time we face this issue - [Support inter-Pod affinity to one or more Pods #68701](https://github.com/kubernetes/kubernetes/issues/68701) is a pretty similar one.

All constraints are evaluated with a `AND` semantics. But even the `AND` can be ambiguous:

1. all constraints must be matched on the same pod
2. all constraints must be matched, but can match on different pods

Current implementation follows the 1st explanation, which is simpler and less error-prone. In the alpha phase, IMO it's good to start with 1st implementation which is also consistent with current PodAffinity semantics.
But I do respect endusers' opinion on this. If users have strong preference on the 2nd, we can update in the next release cycle.

cc/ @krmayankk @bsalamat

Reply:
I think option 1 is counter intuitive, and I would even argue wrong. In the following setup for example, pod 2 has no impact on the skew calculation when scheduling pod1, but pod1 impacts the scheduling of pod2 if it were to be scheduled first. 

If we want simplify the feature and go with option 1, then matchLabels should be defined outside the topologySpreadConstraint.

Pod 1 spec:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: bar
```

Pod 2 spec:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
```

Reply:
I figured I should share the analysis I made of the performance implications of supporting option 2.

This is the current implementation in pseudocode.
```
For each Node (parallel)
  For each Pod
    Match all labels or continue
    For each constraint
      map[topologyKey, value].append(Pod)

For each (topologyKey, value) in map:
  minMap[topologyKey] = min(minMap[topologyKey, len(map[topologyKey, value]))
```

If we want to support counts per label, the algorithm (and complexity) doesn't need to change much.
We move the label match inside the constraints loop. And we keep a set of counters (and minSkew) for each constraint. It can be indexed in a slice. Something like: `[]map[string]int`, where the first index is the number of constraint, then the values of the topologyKey, and then we just have the counter.

In conclusion, the performance is the same. Note that this option also suppresses the need to validate that constraints target different topology keys (which was part of the API validation PR). I think we should do this change sooner rather than later, because it is user facing.

Reply:
> Note that this option also suppresses the need to validate that constraints target different topology keys (which was part of the API validation PR).

The API validation PR is different - that validates the same tuple {topologyKey, scheduleAction} must be unique.

Reply:
> The API validation PR is different - that validates the same tuple {topologyKey, scheduleAction} must be unique.

We are saying that counters need to be per {topologyKey, action, labels}. So it is valid to have different constraints with the same {topologyKey, action}. And at that point, we can simply allow repetitions in general. Having more than one constrain with the same triple is unnecessary for a user, but it's semantically correct.

In the example from @ahg-g, having the same topologyKey for both constraints makes sense under this interpretation.

Reply:
> Having more than one constrain with the same triple is unnecessary for a user, but it's semantically correct.

We should give a validation error to save calculation effort in scheduler side. For the following case:

```yaml
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: bar
```

It's easier to understand with a "merged" yaml like:

```yaml
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchExpressions:
        key: app
        operator: In
        values: ["app", "bar"]
```

Reply:
Good point. You could theoretically have different maxSkew for each label. But I don't think that's a useful use-case. Keeping the validation SGTM.

Reply:
If we have a few constraints, the label selectors of each constraint may or may not match a pod. The constraints whose labels do not match a pod are considered satisfied for that pod. Let's use @ahg-g example:

incoming pod:
```
spec:
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/zone
    selector:
      matchLabels:
        app: foo
  topologySpreadConstraint:
    maxSkew: 1
    topologyKey: k8s.io/region
    selector:
      matchLabels:
        app: bar
```

existing pod:
```
labels:
   app: foo
```

The existing pod matches only the first constraint of the incoming pod. So, we count the existing pod towards the pods in its corresponding `zone`. The existing pod does not affect the count that we keep for each `region` (second constraints) when we check `maxSkew`.

Reply:
That is what I am suggesting that the behavior should be as well, but I don't think the implementation actually behaves this way. May be I am missing something, but it seems that in line [243](https://github.com/kubernetes/kubernetes/blob/d5e205af21ec42f98f0e8b41f3db105e9c566473/pkg/scheduler/algorithm/predicates/metadata.go#L243) we check if the existing pod matches all the constraints, and only then we add it to the minMatch of all the toplogyKeys included in the constraints. So in the example, the existing pod will not be even be counted towards the pods in the corresponding zone.

Reply:
@ahg-g You read it right on current implementation.

Ok, as a majority of people prefer option 2, I will make a commit in this PR to address this.

Thanks @bsalamat @ahg-g! and @alculquicondor for bringing this up!

Reply:
Logic updated as an additional commit. PTAL.

BTW: I will update downstream PRs as well.